### PR TITLE
Improved `swap_linear_with_float8_linear`

### PR DESCRIPTION
**Pull request by awgu**
This PR improves `swap_linear_with_float8_linear` to fix https://github.com/pytorch-labs/float8_experimental/issues/205 and https://github.com/pytorch-labs/float8_experimental/issues/206.

**Test Plan**
Added unit tests for `swap_linear_with_float8_linear` to `test_base.py`.

```
./test/test_fsdp.sh
pytest test/test_base.py
```

### CODE REVIEW

File: float8_experimental/float8_linear_utils.py [hunk 1]
```diff
@@ -66,32 +68,53 @@ def _update_history_with_new_amax(new_amax, amax_history):
...
-    module,
-    emulate=False,
-    skip_fqn_list=None,
-    cur_fqn="",
-):
+    module: nn.Module,
+    module_cls: Type[nn.Module],
+    emulate: bool = False,
+    skip_fqn_list: Optional[List[str]] = None,
+    cur_fqn: str = "",
```
Comment (awgu) [inline float8_experimental/float8_linear_utils.py:75]: @drisspg Do you have historical context on `cur_fqn`?

In particular, did users want to be able to pass `skip_fqn_list` as module FQNs that are "globally" prefixed from the root module, while only call `swap_linear_with_float8_linear` on a submodule?
```
Example:
skip_fqn_list = ["dense_arch.mlp.lin1", "dense_arch.mlp.lin2"]
swap_linear_with_float8_linear(model.dense_arch, ..., skip_fqn_list, cur_fqn="dense_arch")
```

Or, was `cur_fqn` always expected to be `""` upon the user's call to `swap_linear_with_float8_linear`?

If it is the latter case, then personally I would prefer removing this argument as to not leak arguments used for recursion to the user.

**Review by awgu — COMMENTED** [review]
@drisspg Do you know how to run the existing tests?

Following the README with `./test/test_fsdp.py`, I get:
```
    from float8_experimental.float8_linear import Float8Linear
ModuleNotFoundError: No module named 'float8_experimental'
```

I try to use `python -m pytest` to work around these, but some of these tests like `test_fsdp_compile.py` use fire.

**Comment by drisspg** [comment]
For the first question: @awgu  did you install the float8 exp -> `pip install -e .`

**Comment by awgu** [comment]
@drisspg Amazing! Thank you! `pip install -e .` fixed it. (I should have read the README better 😢 )

File: float8_experimental/float8_linear_utils.py [hunk 2]
```diff
@@ -66,32 +68,55 @@ def _update_history_with_new_amax(new_amax, amax_history):
...
-            child, torch.nn.Linear
-        ):
-            new_child = module.from_float(child, emulate)
-            setattr(model, name, new_child)
-        else:
-            swap_linear_with_float8_linear(child, module, emulate)
+    module_names_to_skip = set(skip_fqn_list or [])
+    if isinstance(module, nn.Linear):
+        if cur_fqn not in module_names_to_skip:
+            assert len(list(module.children())) == 0, f"{module}"
```
Comment (drisspg) [inline float8_experimental/float8_linear_utils.py:93]: Maybe a better error message

**Comment by drisspg** [comment]
One follow up I had been thinking about was to skip linear modules that don't have the correct alignment constraints, k%16 ==0

**Pull request closed without merging**